keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,594 @@
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move in the 1.18 cycle this means that not
+all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there could be a minimum version
+check in the presubmit.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
+
+The post-submit job checks out master in k/kubeadm, creates a new branch
+release-x.yy and pushes the branch.
+
+###### k/k ‘s release-x.yy branch is fast-forwarded to master
+
+Currently this is a manual step performed by someone on the release team.
+It might be something that the k/kubeadm maintainers have to do manually
+on a daily basis. Automation is possible but is tricky and needs further
+discussion.
+
+###### k/k pushes a new tag
+
+The post-submit picks the change and based on the tag value,
+checks out the according branch, tags the latest commit in the branch
+and pushes the tag.
+
+##### Picking a commit to build kubeadm from
+
+There is a chicken and egg problem where a tag in the k/kubeadm repository
+cannot be added before the k8s release is cut. This means that the build
+tools in k/release must not depend on looking for a tag in the k/kubeadm
+repository when building kubeadm.
+
+The proposed solution is to always pick the latest commit in a certain
+branch when building.
+
+##### k/release starts building kubeadm version x.yy.
+
+The k/release tooling pulls out the k/kubeadm repo and checks out
+branch “release-x.yy” k/kubeadm build tooling must write a KUBE_VERSION
+file with static versions, but a dynamic build date/time.
+Kubeadm’s version information in the binary should be based on the
+KUBE_VERSION file. The build tools should pick the latest commit of
+“release-x.yy” for the build.
```

Comment:
Is it worth it to also add a "pre-version" tag to identify the commit that was selected so we have a clear pointer for the post-submit job and the following can be relaxed a bit?

>  It must be the responsibility of the kubeadm maintainers to not backport
any last minute changes to the current development “release-x.yy” branch.
This is already a practice while kubeadm resides in k/k.

Reply:
this is not so easy to manage. there is high chance if we forget to manually apply such a tag.
also the k/k automation *does not know which* commit to tag.

i think the only solution is to tag latest commit.

Reply:
Oh yeah, sorry let me explain this better, I'm talking about  the release tooling tagging the latest commit with a `pre-version` tag so that the post-submit can know exactly which commit was used by the release tooling instead of using the latest commit (that might have changed if someone would merge something on the "release-x.yy" branch by error.

Full process would be:

```
1. release tooling checkout "release-x.yy" branch
2. release tooling write a KUBE_VERSION file with the static version 
3. release tooling tags the commit with a "pre-${version}" tag
4. build tool will picks the commit and build it
5. k/k is tagged
6. post-summit job looks for the "pre-${version}" tag on k/kubeadm
7. post-summit job removes the "pre-${version}" tag and adds the "${version}" tag
```

Hope this is a bit clearer, sorry if it wasn't before.

Reply:
ok, this can work.
having the intermediate tag can help both the post-submit and developers.

my concern is the more automation we plan in this KEP, the lower the chances it will be merged as "implementable" and the work will begin. i see some of the automation aspects as things we can improve over time.

Reply:
i have applied this proposal. thanks @LucaLanziani